### PR TITLE
AP-8750 # Added SMS MFA

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: src/apps/services/AWSCognitoClient.ts [hunk 1]
```diff
@@ -501,7 +836,32 @@ export default class AWSCognitoClient {
     )
   }
 
-  async setupMfa() {
+  async setupEmailMfa() {
```
Comment (divporter) [inline src/apps/services/AWSCognitoClient.ts]: 50% a question for myself, when would be setting up email MFA?

Comment (mymattcarroll) [reply src/apps/services/AWSCognitoClient.ts]: code i wrote last week when the requirement was email (not sms) and figured leaving it wont hurt. Happy to remove if you would prefer.

Comment (divporter) [reply src/apps/services/AWSCognitoClient.ts]: Yeah ok, I had a feeling that might be the case.  I'd be more comfortable with it removed as it's not supported. Bit if you're confident there's no risk, then your call.

Comment (mymattcarroll) [reply src/apps/services/AWSCognitoClient.ts]: Removing.

Comment (kizaonline) [reply src/apps/services/AWSCognitoClient.ts]: Zombie code, I think we should remove it

File: src/components/mfa/MfaPhoneNumberDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                    : 'Send again'}
+                </Button>
+              </Box>
+            </>
+          ) : (
+            <>
+              <Typography variant="body2" paragraph>
+                {pendingSmsSetup
+                  ? 'Enter your phone number to receive SMS verification codes when signing in.'
+                  : 'Enter your phone number to save it to your account.'}
```
Comment (divporter) [inline src/components/mfa/MfaPhoneNumberDialog.tsx]: Dumb question, why would a user be saving a phone number not for MFA?

File: src/components/mfa/MultiFactorAuthentication.tsx [hunk 3]
```diff
@@ -1,199 +1,169 @@
...
+        isSettingUp={isSettingUpMfa && settingUpMfaMethod === 'sms'}
+        isSettingPreferredMfaMethod={isSettingPreferredMfaMethod}
+        isSetupDisabled={!!loadingError || isSettingUpMfa}
+        showSetupErrorTooltip={!!loadingError}
+        title="SMS"
+        description="Receive a one-time verification code via SMS each time MFA is required."
+        detail={phoneDetail}
+        cypressPrefix="mfa-sms"
+        onSetup={() => beginMfaSetup('sms')}
+        onDisable={() => beginDisablingMfaMethod('sms')}
```
Comment (divporter) [inline src/components/mfa/MultiFactorAuthentication.tsx:99]: just to draw a distinction from authenticator, so I could disabled SMS MFA, and then re-enable without having to re-enter my phone number?

Comment (mymattcarroll) [reply src/components/mfa/MultiFactorAuthentication.tsx:99]: correct

File: src/components/mfa/MfaPhoneNumberDialog.tsx (hunk 2 above)
Comment (mymattcarroll) [reply src/components/mfa/MfaPhoneNumberDialog.tsx]: checking AI slop...please hold

File: src/hooks/useMfa.tsx [hunk 4]
```diff
@@ -126,86 +252,451 @@ export function MfaProvider({
...
+        isSettingPreferredMfaMethod: true,
+        setupError: undefined,
+      }))
+
+      try {
+        await authService.setPreferredMfaMethod(mfaMethod)
+        setState((currentState) => ({
+          ...currentState,
+          isSettingPreferredMfaMethod: false,
+          mfaSettings: setPreferredMfaMethodInSettings(
```
Comment (divporter) [inline src/hooks/useMfa.tsx:605]: no issue with the function itself, just in this context it looks like a useState setter. I also have no suggestion for a different name

File: src/components/mfa/MfaPhoneNumberDialog.tsx (hunk 2 above)
Comment (mymattcarroll) [reply src/components/mfa/MfaPhoneNumberDialog.tsx]: I asked the AI to future proof it a little bit so that the dialog would be used as a profile editor as well as MFA setup. Happy to remove and make this dialog MFA specific if you'd prefer.

File: src/components/mfa/MfaAuthenticatorAppDialog.tsx [hunk 5]
```diff
@@ -10,25 +10,18 @@ import {
   DialogContent,
   DialogActions,
   Button,
+  DialogTitle,
```
Comment (divporter) [inline src/components/mfa/MfaAuthenticatorAppDialog.tsx:13]: Could I be a butt and request renaming this file to MfaAuthenticatorDialog or some such? On first review I thought it was covering both Authenticator and SMS.

File: src/components/mfa/MfaPhoneNumberDialog.tsx (hunk 2 above)
Comment (mymattcarroll) [reply src/components/mfa/MfaPhoneNumberDialog.tsx]: Removed

File: src/components/mfa/MfaPhoneNumberDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        }
+      }}
+      fullWidth
+      maxWidth="sm"
+    >
+      <form onSubmit={handleSubmit}>
+        <DialogTitle>
+          {isPhoneVerificationRequired
+            ? 'Verify Phone Number'
+            : 'Save Phone Number'}
```
Comment (divporter) [inline src/components/mfa/MfaPhoneNumberDialog.tsx:160]: are these and the submit buttons still relevant now, or should it only say Verify Phone Number?

Comment (mymattcarroll) [reply src/components/mfa/MfaPhoneNumberDialog.tsx:160]: They are relevant. User saves phone number first, that enters code sent via SMS to verify phone number.

**Comment by mymattcarroll** [comment]
Thorough review! Thank you @divporter.
